### PR TITLE
[TEST] Add a unittest for adding "FOSSUnitedProfile"

### PR DESCRIPTION
**Pull request by rahulporuri**
## What type of PR is this? (check all applicable)
- [ ] 🍕Feature
- [ ] ⚙️Chore
- [ ] 🐛Bug Fix
- [ ] 🎨Style
- [ ] 🧑‍💻Refactor
- [ ] 📕Documentation
- [ ] 🏎️Optimization
- [x] Testing

## Description

This PR adds a unittest that tests the creation of a `FOSSUnitedProfile`. The unittest is written as follows

- check that the profile that we are interested in adding does not already exist
- add and insert a Frappe `User` and `FOSSUnitedProfile`. Note that this is because `FOSSUnitedProfile` requires the presence of a `User`
- verify that the inserted `FOSSUnitedProfile` was actually added by querying for it

Unittests usually follow the format "Given, When, Then" e.g. Given a scenario, When an action is performed, Then verify the expected outcomes.

The unittest runs on CI as expected

```
----------------------------------------------------------------------
Ran 1 test in 0.008s

OK
```

## Related Issues & Docs
N/A

## Checklist
- [ ] ~I have read the [contribution guidelines]("./docs/CONTRIBUTING.md").~
- [x] I have tested these changes locally.
- [ ] ~I have updated the documentation (If applicable).~
- [x] The code follows the project's coding standards.
- [x] I have added/updated tests, if applicable.

## Screenshots/GIFs/Screen Recordings (if applicable)
N/A

## Additional context
N/A

## [optional] What gif best describes this PR or how it makes you feel?
N/A

### CODE REVIEW

File: fossunited/foss_profiles/doctype/foss_user_profile/test_foss_user_profile.py [hunk 1]
```diff
@@ -1,9 +1,52 @@
...
+        profile = frappe.get_doc(
+            {
+                "doctype": "FOSS User Profile",
+                "user": frappe_user.name,
+                "username": frappe_user.username,
+                "is_published": 1,
+            },
+        )
+        profile.db_insert()
+        inserted_name = profile.name
```
Comment (NagariaHussain) [inline fossunited/foss_profiles/doctype/foss_user_profile/test_foss_user_profile.py:45]: Why db insert instead of direct insert call?

**Review by NagariaHussain — CHANGES_REQUESTED** [review]

File: fossunited/foss_profiles/doctype/foss_user_profile/test_foss_user_profile.py [hunk 2]
```diff
@@ -1,9 +1,52 @@
...
+        frappe_user.db_insert()
+        profile = frappe.get_doc(
+            {
+                "doctype": "FOSS User Profile",
+                "user": frappe_user.name,
+                "username": frappe_user.username,
+                "is_published": 1,
+            },
+        )
+        profile.db_insert()
```
Comment (NagariaHussain) [inline fossunited/foss_profiles/doctype/foss_user_profile/test_foss_user_profile.py]:
```suggestion
        profile = frappe.get_doc(
            {
                "doctype": "FOSS User Profile",
                "user": frappe_user.name,
                "username": frappe_user.username,
                "is_published": 1,
            },
        ).insert()
```

File: fossunited/foss_profiles/doctype/foss_user_profile/test_foss_user_profile.py [hunk 3]
```diff
@@ -1,9 +1,52 @@
...
+                "doctype": "User",
+                "username": inserted_username,
+                # Create a unique email address as it is used as a database key
+                "email": str(uuid.uuid4()) + "@fossunited.org",
+                "first_name": "wish",
+                "name": inserted_name,
+                "full_name": inserted_name,
+            },
+        )
+        frappe_user.db_insert()
```
Comment (NagariaHussain) [inline fossunited/foss_profiles/doctype/foss_user_profile/test_foss_user_profile.py]:
```suggestion
        frappe_user = frappe.get_doc(
            {
                "doctype": "User",
                "username": inserted_username,
                # Create a unique email address as it is used as a database key
                "email": str(uuid.uuid4()) + "@fossunited.org",
                "first_name": "wish",
                "name": inserted_name,
                "full_name": inserted_name,
            },
        ).insert()
```

File: fossunited/foss_profiles/doctype/foss_user_profile/test_foss_user_profile.py [hunk 4]
```diff
@@ -1,9 +1,52 @@
...
+        )
+        profile.db_insert()
+        inserted_name = profile.name
+
+        # Then verify that the Profile has been stored as expected
+        profiles_metadata = frappe.get_all(
+            doctype="FOSS User Profile",
+            filters={"username": inserted_username},
+        )
+        self.assertEqual(profiles_metadata, [{"name": inserted_name}])
```
Comment (NagariaHussain) [inline fossunited/foss_profiles/doctype/foss_user_profile/test_foss_user_profile.py]:
```suggestion
        profile_exists = frappe.db.exists(
            "FOSS User Profile",
            {"username": inserted_username}
        )
        self.assertTrue(profile_exists)
```

**Review by NagariaHussain — DISMISSED** [review]

File: fossunited/foss_profiles/doctype/foss_user_profile/test_foss_user_profile.py [hunk 5]
```diff
@@ -1,8 +1,50 @@
...
-    pass
+    def test_add_profile(self):
+        # Given a user that does not have a FOSSUnitedProfile
+        inserted_username = "wisharya"
+        inserted_name = "wish arya"
+        profiles_metadata = frappe.get_all(
+            doctype="FOSS User Profile",
+            filters={"username": inserted_username},
+        )
+        self.assertEqual(profiles_metadata, [])
```
Comment (rahulporuri) [inline fossunited/foss_profiles/doctype/foss_user_profile/test_foss_user_profile.py]:
```suggestion        
        profile_exists = frappe.db.exists(
            "FOSS User Profile",
            {"username": inserted_username}
        )
        self.assertFalse(profile_exists)
```

File: fossunited/foss_profiles/doctype/foss_user_profile/test_foss_user_profile.py [hunk 6]
```diff
@@ -1,8 +1,50 @@
...
+        ).insert()
+        profile = frappe.get_doc(
+            {
+                "doctype": "FOSS User Profile",
+                "user": frappe_user.name,
+                "username": frappe_user.username,
+                "is_published": 1,
+            },
+        ).insert()
+        inserted_name = profile.name
```
Comment (rahulporuri) [inline fossunited/foss_profiles/doctype/foss_user_profile/test_foss_user_profile.py]:
```suggestion
```

File: fossunited/foss_profiles/doctype/foss_user_profile/test_foss_user_profile.py [hunk 7]
```diff
@@ -1,8 +1,49 @@
...
 class TestFOSSUserProfile(FrappeTestCase):
-    pass
+    def test_add_profile(self):
+        # Given a user that does not have a FOSSUnitedProfile
+        inserted_username = "wisharya"
+        inserted_name = "wish arya"
+        profile_exists = frappe.db.exists(
+            "FOSS User Profile",
+            {"username": inserted_username}
+        )
```
Comment (rahulporuri) [inline fossunited/foss_profiles/doctype/foss_user_profile/test_foss_user_profile.py]:
```suggestion
        profile_exists = frappe.db.exists(
            "FOSS User Profile", {"username": inserted_username}
        )
```

File: fossunited/foss_profiles/doctype/foss_user_profile/test_foss_user_profile.py [hunk 8]
```diff
@@ -1,8 +1,49 @@
...
+                "user": frappe_user.name,
+                "username": frappe_user.username,
+                "is_published": 1,
+            },
+        ).insert()
+
+        # Then verify that the Profile has been stored as expected
+        profile_exists = frappe.db.exists(
+            "FOSS User Profile",
+            {"username": inserted_username}
```
Comment (rahulporuri) [inline fossunited/foss_profiles/doctype/foss_user_profile/test_foss_user_profile.py]:
```suggestion
            "FOSS User Profile", {"username": inserted_username}
```

File: fossunited/foss_profiles/doctype/foss_user_profile/test_foss_user_profile.py [hunk 9]
```diff
@@ -1,8 +1,47 @@
...
+            {
+                "doctype": "User",
+                "username": inserted_username,
+                # Create a unique email address as it is used as a database key
+                "email": str(uuid.uuid4()) + "@fossunited.org",
+                "first_name": "wish",
+                "name": inserted_name,
+                "full_name": inserted_name,
+            },
+        ).insert()
```
Comment (harshtandiya) [inline fossunited/foss_profiles/doctype/foss_user_profile/test_foss_user_profile.py:33]: This step must be creating a foss user on it's own. 
Ref: [user_utils.py file](https://github.com/fossunited/fossunited/blob/develop/fossunited/fossunited/user_utils.py) , function `create_profile_on_user_create`
Also
[Hooks file](https://github.com/fossunited/fossunited/blob/develop/fossunited/hooks.py) line 181.


This is why tests are failing on next line with duplicate entry error. A profile already exists with this user linked to it.

Suggestion: don't do line 34 to 41, instead just check if profile exists that is linked to the user.

```py
profile_exists = frappe.db.exists(
    "FOSS User Profile", {"user": frappe_user.name}
)
```

Comment (rahulporuri) [reply fossunited/foss_profiles/doctype/foss_user_profile/test_foss_user_profile.py:33]: perfect. thanks for debugging this @harshtandiya ! I can only make the changes on Friday now. @Mangeshrex @harshtandiya  any chance you could pick this up and address it tomorrow? If not, I will address it on Friday.

Comment (mngshm) [reply fossunited/foss_profiles/doctype/foss_user_profile/test_foss_user_profile.py:33]: The tests seems to pass when I'm trying to do what harsh has suggested. 
Capture-Jul-11-18-54-01

**Comment by rahulporuri** [comment]
@harshtandiya you will have to review and approve this PR now. @harshtandiya and I are contributors :D

**Review by harshtandiya — APPROVED** [review]

**Pull request merged**